### ISSUE TITLE
Issues with container image generation workflow

### ISSUE THREAD

**Issue by andreabolognani** [kubevirt/libvirt#33]
Disclaimer: I have only cursory understanding of Docker, and I'm mostly unfamiliar with KubeVirt's development workflow, so take all everything below with a grain of salt and correct me when I'm wrong!

While looking into the way the `kubevirt/libvirt` container image is generated, I have found a number of issues. Some of them are cosmetic or relatively innocuous:

* `README.md` is out of date: it tells you to use an image which is not the same as KubeVirt is using. Since keeping the two aligned is clearly error prone, perhaps just pointing to the corresponding information in `kubevirt/kubevirt` would be better?

* The `build.sh` and `publish.sh` scripts seem to be entirely redundant, given that the intended workflow (at least as far as I can tell) expects container images to be published automatically as part of a successful build on Travis CI. Do they actually serve a purpose, or can they be dropped?

* The scripts mentioned above also happen to hardcode (multiple times) the expected libvirt version, which is already present in the `Dockerfile`: if they need to stick around, can we figure out a way for the libvirt version to be specified in a single place and then make everything else consume it from there?

* Speaking of versioning, right now the container images are versioned the same as the libvirt RPMs installed inside them, but that doesn't seem enough to me: what if the images are rebuilt after `fedora:30` has been updated, or more simply when the `Dockerfile` has been changed to eg. add or remove packages? I think the tag should contain, at the very least, a shortened version of the git commit hash, making it unique; ideally we'd also have a version number (single digit would do) for user convenience, but guaranteeing that's always bumped as intended is trickier so perhaps that could be a follow-up enhancement.

* The `latest` tag seems to be pointing to a random outdated image (`4.1.0`) and should either be moved or, and this is arguably a better option, dropped altogether.

There's a more fundamental issue, though: due to the way container images are currently being generated, both the `x86_64` and the `ppc64le` builds contend for the same tag name, with the one that finishes last winning. Looking at

  https://hub.docker.com/r/kubevirt/libvirt/tags

you can see that some of the images are only offered for `x86_64` and some are only available for `ppc64le`; what we actually want is something like

  https://hub.docker.com/r/andreabolognani/kubevirt-libvirt/tags

instead, where the same tag can be pulled on either architecture and it will work just fine, as in

```
$ docker run --rm -it \
  andreabolognani/kubevirt-libvirt:5.0.0-2.fc30 \
  /bin/sh -c 'uname -m && libvirtd --version'
x86_64
libvirtd (libvirt) 5.0.0
$
```

and

```
$ docker run --rm -it \
  andreabolognani/kubevirt-libvirt:5.0.0-2.fc30 \
  /bin/sh -c 'uname -m && libvirtd --version'
ppc64le
libvirtd (libvirt) 5.0.0
$
```

To obtain the result above, what I did is build the image on both `x86_64` and `ppc64le` from the unmodified `Dockerfile`, tagging it as `5.0.0-2.fc30.x86_64` and `5.0.0-2.fc30.ppc64le` respectively, then pushed both to Docker Hub; then I pulled both on one of the two machines (this last step might have been unnecessary) and executed

```
$ docker manifest create \
  andreabolognani/kubevirt-libvirt:5.0.0-2.fc30 \
  andreabolognani/kubevirt-libvirt:5.0.0-2.fc30.x86_64 \
  andreabolognani/kubevirt-libvirt:5.0.0-2.fc30.ppc64le
$ docker manifest push \
  andreabolognani/kubevirt-libvirt:5.0.0-2.fc30
$
```

after which I manually deleted the architecture-specific tags using the Web interface.

Note that the last two steps will not work on Quay because it apparently doesn't yet support the newest manifest format, but as far as Docker Hub is concerned everything is fine.

Now the question is, how to automate the last steps? One silly idea would be to have one of the two build workers stick around until both architecture-specific tags have appeared on Docker Hub, at which point it would create and push the manifest. Any better ideas?

I also haven't found a way to delete remote tags from the command line, so perhaps that bit of cleanup will have to happen manually?

Any thoughts?

**Comment by andreabolognani** [comment]
> To obtain the result above, what I did is build the image on both `x86_64` and `ppc64le` from the unmodified `Dockerfile`, tagging it as `5.0.0-2.fc30.x86_64` and `5.0.0-2.fc30.ppc64le` respectively, then pushed both to Docker Hub; then I pulled both on one of the two machines (this last step might have been unnecessary)

Based on further testing, it looks like pulling the second image is indeed not necessary.

> I also haven't found a way to delete remote tags from the command line, so perhaps that bit of cleanup will have to happen manually?

Turns out this can be done, tough it's a bit clunky: assuming neither `$DOCKER_USER` nor `$DOCKER_PASS` need to be escaped to be valid JSON strings, then something like

```
ORGANIZATION=kubevirt
IMAGE=libvirt
TAG=5.0.0-2.fc30.ppc64le

TOKEN=$(curl -s -X POST -H "Content-Type: application/json" -d "{\"username\": \"$DOCKER_USER\", \"password\": \"$DOCKER_PASS\"}" https://hub.docker.com/v2/users/login/ | jq -r .token)

curl -X DELETE -H "Authorization: JWT ${TOKEN}" "https://hub.docker.com/v2/repositories/${ORGANIZATION}/${IMAGE}/tags/${TAG}/"
```

(adapted from [this blog post](https://devopsheaven.com/docker/dockerhub/2018/04/09/delete-docker-image-tag-dockerhub.html)) will do the trick.

**Comment by slintes** [comment]
> * The `build.sh` and `publish.sh` scripts seem to be entirely redundant, given that the intended workflow (at least as far as I can tell) expects container images to be published automatically as part of a successful build on Travis CI. Do they actually serve a purpose, or can they be dropped?

`build.sh` is called by the script part in Travis, which is executed for every PR, to catch problems.
`publish.sh` is called by the deploy part in Travis, which executed on master only, so after PRs have been merged.
So it makes sense to keep both.

> * ... can we figure out a way for the libvirt version to be specified in a single place and then make everything else consume it from there?

That would be great! :wink: 

> * Speaking of versioning, right now the container images are versioned the same as the libvirt RPMs installed inside them, but that doesn't seem enough to me: what if the images are rebuilt after `fedora:30` has been updated, or more simply when the `Dockerfile` has been changed to eg. add or remove packages? I think the tag should contain, at the very least, a shortened version of the git commit hash, making it unique; ideally we'd also have a version number (single digit would do) for user convenience, but guaranteeing that's always bumped as intended is trickier so perhaps that could be a follow-up enhancement.

That's basically why we reference images by shasum, and not by tag when consuming them. The shasum never changes, you always get the exact same image. So it's probably not worth the effort to improve tag names (at least as long as we stay on dockerhub, on quay.io images without tags are deleted soonish AFAIK)

> * The `latest` tag seems to be pointing to a random outdated image (`4.1.0`) and should either be moved or, and this is arguably a better option, dropped altogether.

It's a bad but common practice to use latest. So I personally would prefer to always push the latest tag on master alongside with the versioned tag, instead of removing it. Thanks for raising this, the current situation is bad for sure! 

The multiarch issue must be resolved as well, good catch! But I have no experience with this.

/cc @rmohr 
/cc @cynepco3hahue

**Comment by andreabolognani** [comment]
> > * The `build.sh` and `publish.sh` scripts seem to be entirely redundant, given that the intended workflow (at least as far as I can tell) expects container images to be published automatically as part of a successful build on Travis CI. Do they actually serve a purpose, or can they be dropped?
> 
> `build.sh` is called by the script part in Travis, which is executed for every PR, to catch problems.
> `publish.sh` is called by the deploy part in Travis, which executed on master only, so after PRs have been merged.
> So it makes sense to keep both.

Are they, though?

```
$ git grep -E 'build|publish'
.travis.yml:- docker build . -t kubevirt/libvirt
build.sh:docker build -t kubevirt/libvirt:5.0.0 .
$
```

Perhaps that was the intention, but it certainly doesn't seem to be the case as things are today.

> > * ... can we figure out a way for the libvirt version to be specified in a single place and then make everything else consume it from there?
> 
> That would be great! wink

I didn't say that explicitly, but I can spend time attacking at least some of these issues after we've reached an agreement on what to do :)

> > * Speaking of versioning, right now the container images are versioned the same as the libvirt RPMs installed inside them, but that doesn't seem enough to me: what if the images are rebuilt after `fedora:30` has been updated, or more simply when the `Dockerfile` has been changed to eg. add or remove packages? I think the tag should contain, at the very least, a shortened version of the git commit hash, making it unique; ideally we'd also have a version number (single digit would do) for user convenience, but guaranteeing that's always bumped as intended is trickier so perhaps that could be a follow-up enhancement.
> 
> That's basically why we reference images by shasum, and not by tag when consuming them. The shasum never changes, you always get the exact same image. So it's probably not worth the effort to improve tag names (at least as long as we stay on dockerhub, on quay.io images without tags are deleted soonish AFAIK)

I'm probably missing something obvious, but how can you pull the image by shasum? The obvious way don't seem to work:

```
$ docker pull 840d4502f48f567f9030e22ba8aa8294003f1ac20b6a0c06c01b0236a8964a3a
invalid repository name (840d4502f48f567f9030e22ba8aa8294003f1ac20b6a0c06c01b0236a8964a3a), cannot specify 64-byte hexadecimal strings
$ docker pull sha256:840d4502f48f567f9030e22ba8aa8294003f1ac20b6a0c06c01b0236a8964a3a
Error response from daemon: pull access denied for sha256, repository does not exist or may require 'docker login'
$
```

Either way, including the short commit hash (and thus avoiding the issue altogether) seems simple enough, so unless there's strong opposition to the idea I think it makes sense to do that.

> > * The `latest` tag seems to be pointing to a random outdated image (`4.1.0`) and should either be moved or, and this is arguably a better option, dropped altogether.
> 
> It's a bad but common practice to use latest. So I personally would prefer to always push the latest tag on master alongside with the versioned tag, instead of removing it.

Personally I would favor dropping it, but either solution is better than the current status quo :)

> Thanks for raising this, the current situation is bad for sure!
> 
> The multiarch issue must be resolved as well, good catch! But I have no experience with this.

I have very little experience too... All of the information I presented is from playing around with it this week. Hopefully more knowledgeable people will jump in, but if not I think I can probably cobble something reasonable together with enough effort.

**Comment by rmohr** [comment]
@andreabolognani great to see you on the repo :)

> Are they, though?

Yep, you are right they are not. Seems like we only use `deploy.sh`.

> I'm probably missing something obvious, but how can you pull the image by shasum? The obvious way don't seem to work:

Here is a good description:

https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier

> Either way, including the short commit hash (and thus avoiding the issue altogether) seems simple enough, so unless there's strong opposition to the idea I think it makes sense to do that.

Having the commit-sha as part of the tag sounds great in general. How does it solve the problem of changing RPMs?

> I have very little experience too... All of the information I presented is from playing around with it this week. Hopefully more knowledgeable people will jump in, but if not I think I can probably cobble something reasonable together with enough effort.

I think that our main problem is that while travis supports pipelines, the pipeline support is limited to one build/arch, so creating a true multi-arch manifest is difficult with it. But we can at least build them with individuall suffixes, like  `*-linux-amd64`, `*-linux-ppc64`, ...

**Comment by andreabolognani** [comment]
> @andreabolognani great to see you on the repo :)

:)

> > Are they, though?
> 
> Yep, you are right they are not. Seems like we only use `deploy.sh`.

So, do we want to rework `.travis.yml` and `deploy.sh` so that they use `build.sh` and `publish.sh`, or ditch the unused scripts? The former probably makes more sense if we're going to make the deploy step more complicated (see below).

> > I'm probably missing something obvious, but how can you pull the image by shasum? The obvious way don't seem to work:
> 
> Here is a good description:
> 
> https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier

Should have spent a couple more minutes in a search engine O:-) Thanks for the pointer!

> > Either way, including the short commit hash (and thus avoiding the issue altogether) seems simple enough, so unless there's strong opposition to the idea I think it makes sense to do that.
> 
> Having the commit-sha as part of the tag sounds great in general. How does it solve the problem of changing RPMs?

It does not :) However, while for base images such as `fedora:30` it seems that it's relatively common to update an existing tag, for RPMs each build will usually have a corresponding changelog entry and a bumped version number, so it shouldn't be much of a problem in practice.

> > I have very little experience too... All of the information I presented is from playing around with it this week. Hopefully more knowledgeable people will jump in, but if not I think I can probably cobble something reasonable together with enough effort.
> 
> I think that our main problem is that while travis supports pipelines, the pipeline support is limited to one build/arch, so creating a true multi-arch manifest is difficult with it. But we can at least build them with individuall suffixes, like `*-linux-amd64`, `*-linux-ppc64`, ...

I was hoping for a cleaner solution, but I think my initial idea of making one of the two build workers responsible for the creation of the manifest, while not very pretty, can still work.

Basically both workers would build and push the (appropriately tagged) container image for their architecture, but one of them would also stick around until all builds are done and finally create and push the manifest.

What do you think about that idea? I can try and build a prototype.

Failing that, I agree that having properly tagged images would already be a big step forward.

**Comment by rmohr** [comment]
> > > Either way, including the short commit hash (and thus avoiding the issue altogether) seems simple enough, so unless there's strong opposition to the idea I think it makes sense to do that.
> > 
> > 
> > Having the commit-sha as part of the tag sounds great in general. How does it solve the problem of changing RPMs?
> 
> It does not :) However, while for base images such as `fedora:30` it seems that it's relatively common to update an existing tag, for RPMs each build will usually have a corresponding changelog entry and a bumped version number, so it shouldn't be much of a problem in practice.

:+1: 

> 
> > > I have very little experience too... All of the information I presented is from playing around with it this week. Hopefully more knowledgeable people will jump in, but if not I think I can probably cobble something reasonable together with enough effort.
> > 
> > 
> > I think that our main problem is that while travis supports pipelines, the pipeline support is limited to one build/arch, so creating a true multi-arch manifest is difficult with it. But we can at least build them with individuall suffixes, like `*-linux-amd64`, `*-linux-ppc64`, ...
> 
> I was hoping for a cleaner solution, but I think my initial idea of making one of the two build workers responsible for the creation of the manifest, while not very pretty, can still work.

Sounds like it could work :+1: 

Thanks for doing the cleanup.

**Comment by slintes** [comment]
> > > Are they, though?
> > 
> > 
> > Yep, you are right they are not. Seems like we only use `deploy.sh`.

oh. sorry... somehow my eyes only catched "build" and assumed it is build.sh, ignoring everything around it :laughing:

**Comment by andreabolognani** [comment]
All the issues I had identified have been addressed as of https://github.com/kubevirt/libvirt/commit/aa0cd4c03fd69607d8c5a5e372dfc3acce5b73fb, so I'm closing this issue now :)
